src/main/java/org/apache/maven/plugin/compiler/TestCompilerMojo.java

```diff
@@ -227,6 +229,37 @@ protected void preparePaths( Set<File> sourceFiles )
             }
         }
 
+        // It is possible that we are building a multi-version jar file in which the main classes are not module based.
+        if ( ! mainModuleDescriptorClassFile.exists() && release != null )
+        {
+            int version = Integer.valueOf( release );
+            while ( version >= 9 )
+            {
+                String releaseOutputDirectory = String.format( "%s%sMETA-INF%sversions%s%d",
+                        getProject().getBuild().getOutputDirectory(), File.separator, File.separator,
+                        File.separator, version );
```

Comment:
Please use `File metaInfDir = new File(getProject().getBuild().getOutputDirectory(), "META-INF/versions/" + version)`. It will do the conversion  to system-specific separators for you.